### PR TITLE
[EAK-365] Implemented RequestAttributeInjector; template method refactoring

### PR DESCRIPTION
**Pull request by mykhailIvchenko**
[EAK-365] implemented RequestAttributeInjector; template method refactoring

### CODE REVIEW

**Comment by mykhailIvchenko** [comment]
I have read the CLA Document and I hereby sign the CLA

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        Object value = annotationValueSupplier.get();
+
+        if( value == null) {
+            defaultMessage();
+        }
+
+        return value;
+    }
+
+    public abstract T getAnnotation(AnnotatedElement element);
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java:58]: Consider lowering the visibility. TBD: are we going to make this abstract class a part of he Toolkit's public API (says: "everyone, you are now welcome to inherit your injectors from this!" . Seems unlikely ...) For the injectors of our very project, package private is enough

**Review by smiakchilo — COMMENTED** [review]
Thank you for this contribution!) Generally a good first issue

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestAttributeInjector.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+        SlingHttpServletRequest request = AdaptationUtil.getRequest(adaptable);
+        name = StringUtils.defaultIfEmpty(annotation.name(), name);
+        Object attribute = request.getAttribute(name);
+
+        if (Objects.isNull(attribute)) {
+            return null;
+        }
+
+        result = getByType(attribute, Integer.class, type)
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestAttributeInjector.java:62]: (cosmetic) Something wrong with code formatting here (we don't use "ladder")

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.sling.models.spi.Injector;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Type;
+import java.util.function.Supplier;
+
+public abstract class BaseInjectorTemplateMethod<T extends Annotation, R extends Adaptable> implements Injector {
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java]: corrected. Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+
+    protected T type;
+    protected R resource;
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java]: corrected. 
migrate to youse generic only for Annotation type.

Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    @CheckForNull
+    @Override
+    public Object getValue(
+        @Nonnull Object o,
+        String name,
+        @Nonnull Type type,
+        @Nonnull AnnotatedElement annotatedElement,
+        @Nonnull DisposalCallbackRegistry disposalCallbackRegistry
+    ) {
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java]: corrected. Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        T annotation = getAnnotation(annotatedElement);
+
+        if (annotation == null) {
+            return null;
+        }
+
+        R adaptable = getAdaptable(o);
+
+        if(adaptable == null) {
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java]: corrected. Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestAttributeInjector.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public void defaultMessage() {
+        LOG.debug(InjectorConstants.EXCEPTION_UNSUPPORTED_TYPE, type);
+    }
+
+
+    private Object getValueFromRequest(
+        SlingHttpServletRequest request,
+        String paramName) {
+        return request.getAttribute(paramName);
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestAttributeInjector.java]: method removed.

Thank you

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestParamInjector.java [hunk 8]
```diff
@@ -80,48 +81,61 @@ public Object getValue(
...
+    @Override
+    public RequestParam getAnnotation(AnnotatedElement element) {
+        return element.getDeclaredAnnotation(RequestParam.class);
+    }
+    @Override
+    public SlingHttpServletRequest getAdaptable(Object object) {
+        return AdaptationUtil.getRequest(object);
+    }
+    @Override
+    public Supplier<Object> getAnnotationValueSupplier(Object request, String name, Type type, AnnotatedElement element, DisposalCallbackRegistry disposalCallbackRegistry, RequestParam annotation) {
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestParamInjector.java]: corrected. Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java]: corrected.

responsibility for casting adaptable is on the child classes.

Thank you

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java (hunk 1 above)
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java]: corrected. Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    public abstract Supplier<Object> getAnnotationValueSupplier(
+        Object o,
+        String name,
+        Type type,
+        AnnotatedElement element,
+        DisposalCallbackRegistry disposalCallbackRegistry,
+        T annotation
+    );
+
+    public abstract void defaultMessage();
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java]: Corrected.

Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestAttributeInjector.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        Object request,
+        String name,
+        Type type,
+        AnnotatedElement element,
+        DisposalCallbackRegistry disposalCallbackRegistry,
+        RequestAttribute annotation
+    ) {
+        String finalName = StringUtils.defaultIfEmpty(annotation.name(), name);
+
+        return () -> getValueFromRequest((SlingHttpServletRequest) request,finalName);
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestAttributeInjector.java]: Corrected.

Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java]: Corrected.

Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestParamInjector.java [hunk 11]
```diff
@@ -80,48 +81,61 @@ public Object getValue(
...
 
-        if (TypeUtil.isValidObjectType(type, String.class)) {
-            return request.getParameter(paramName);
+        return () -> {
 
-        } else if (TypeUtil.isValidArray(type, String.class)) {
-            return getFilteredRequestParameters(request, paramName)
-                .map(RequestParameter::getString)
-                .toArray(String[]::new);
+            if (TypeUtil.isValidObjectType(type, String.class)) {
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestParamInjector.java]: Corrected.

Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestParamInjector.java [hunk 12]
```diff
@@ -80,48 +81,61 @@ public Object getValue(
...
-        } else if (TypeUtil.isValidObjectType(type, RequestParameter.class)) {
-            return request.getRequestParameter(paramName);
+            } else if (TypeUtil.isValidCollection(type, String.class)) {
+                return getFilteredRequestParameters((SlingHttpServletRequest) request, paramName)
+                    .map(RequestParameter::getString)
+                    .collect(Collectors.toList());
 
-        } else if (TypeUtil.isValidCollection(type, RequestParameter.class)) {
-            return request.getRequestParameterList();
+            } else if (TypeUtil.isValidObjectType(type, RequestParameter.class)) {
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/RequestParamInjector.java]: boxed and unboxed values support has been added.

Thank you

File: core/src/test/java/com/exadel/aem/toolkit/core/injectors/models/TestModelRequestAttribute.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    private List<String> requestAttributeStringList;
+    @RequestAttribute(name = "booleanList")
+    private List<Boolean> requestAttributeBooleanList;
+    @RequestAttribute(name = "integerList")
+    private List<Integer> requestAttributeIntegerList;
+    @RequestAttribute(name = "longList")
+    private List<Long> requestAttributeLongList;
+
+    @RequestAttribute(name = "stringArray")
+    private String[] requestAttributeStringArray;
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/test/java/com/exadel/aem/toolkit/core/injectors/models/TestModelRequestAttribute.java:33]: boxed and unboxed values support has been added.

Thank you

File: core/src/test/java/com/exadel/aem/toolkit/core/injectors/models/TestModelRequestAttribute.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+import com.exadel.aem.toolkit.api.annotations.injectors.RequestAttribute;
+
+@SuppressWarnings("unused")
+@Model(adaptables = SlingHttpServletRequest.class,
+    defaultInjectionStrategy = DefaultInjectionStrategy.OPTIONAL
+)
+public class TestModelRequestAttribute {
+
+    @RequestAttribute
+    private String param1;
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/test/java/com/exadel/aem/toolkit/core/injectors/models/TestModelRequestAttribute.java]: name has been adapted.

Thank you.

File: core/src/test/java/com/exadel/aem/toolkit/core/injectors/models/TestModelRequestAttribute.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    @RequestAttribute(name = "param1")
+    private String namedParam;
+
+    private Object paramObjectType;
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/test/java/com/exadel/aem/toolkit/core/injectors/models/TestModelRequestAttribute.java]: removed.

Thank you

File: core/src/main/java/com/exadel/aem/toolkit/api/annotations/injectors/RequestAttribute.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package com.exadel.aem.toolkit.api.annotations.injectors;
+
+import org.apache.sling.models.annotations.Source;
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/api/annotations/injectors/RequestAttribute.java:3]: corrected. 

Thank you.

File: core/src/main/java/com/exadel/aem/toolkit/api/annotations/injectors/RequestAttribute.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+import org.apache.sling.models.spi.injectorspecific.InjectAnnotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import com.exadel.aem.toolkit.core.injectors.RequestAttributeInjector;
+
+@Target({ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER})
```
[the remark this answers is not in the document]
Comment (mykhailIvchenko) [reply core/src/main/java/com/exadel/aem/toolkit/api/annotations/injectors/RequestAttribute.java:24]: corrected. 

Thank you.

**Pull request merged**